MAINT: Remove redundant code in MLP

rth opened the pull request:
Closes https://github.com/scikit-learn/scikit-learn/issues/14807
Closes https://github.com/scikit-learn/scikit-learn/issues/11952

Alternative to https://github.com/scikit-learn/scikit-learn/pull/10665


This makes sure that it's possible to use a custom activation function registering new ones in `sklearn.neural_network.multilayer_perceptron.{ACTIVATIONS,DERIVATIVES}`.

It was almost working before, except for a manual check of supported activation (instead of using the `ACTIVATIONS` variable), so this mostly adds a test.

Not sure if we want to document it (and if so where).

rth committed:
Allow custom activation in MLP

rth commented:
A minimal example would be something like,
```py
from sklearn.neural_network.multilayer_perceptron import (
   ACTIVATIONS, DERIVATIVES, MLPClassifier
)                                 

ACTIVATIONS['identity2'] = lambda x: x                                                                                                 

def inplace_identity_derivative2(Z, delta): 
     pass 
                                                                                                                                        
DERIVATIVES['identity2'] = inplace_identity_derivative2                                                                                

mlp = MLPClassifier(activation='identity2')                                                                                            

mlp.fit([[0, 1], [1, 1]], [0, 1])
```

changing public variables in the scikit-learn namespace is not great, but as a somewhat experimental solution it should work.

rth committed:
fmt

jnothman reviewed (approved):
Sure, why not?

adrinjalali commented:
I'm very worried about this one. Which one of the following two are we doing?

- Allow people to use this API, which is _public_. If we're doing this, then we should document it in `neural_networks_supervised.rst` at least, and add a whats_new entry. But that means we're actually happy about this solution; are we? I rather let them pass a function instead of having to add the function to the dictionary.
- We want to have a better solution, and we'll probably do that, but later. That means we would have to deprecate this solution which is just dirty, if we already know we don't like it.

I kinda don't like not documenting it, cause we'll force ourselves to deprecate the behavior if we wanna change it anyway. Which means why not document and let people use it then?

rth commented:
> Allow people to use this API, which is public. If we're doing this, then we should document it in neural_networks_supervised.rst at least, and add a whats_new entry

We don't have to document anything if we don't judge it necessary. The one line change to the MLP file could be done just under the justification of removal of redundant code unless you think it's controversial. It happens that you can then provide custom activation by adding to these dict variables. Is it worth explicitly disallowing that? I don't know. Technically `ACTIVATIONS, DERIVATIVES` are defined in `sklearn.neural_network._base` which is private it's just that I can't monkeypatch it there (that doesn't work).

I can locally import those inside `MLPClassifier` and remove top level imports in `multilayer_perceptron.py` (which would also allow using this mechanism by only changing private variables) if you think that would be preferrable?

>  I rather let them pass a function instead of having to add the function to the dictionary.

We could pass a function that returns the activation function and the derivative. The problem is that activation and derivatives are not computed at the same time (meaning it would be slower) and also add a bit more complexity to code.

adrinjalali commented:
> I can locally import those inside `MLPClassifier` and remove top level imports in `multilayer_perceptron.py` (which would also allow using this mechanism by only changing private variables) if you think that would be preferrable?

I would prefer that.

> We could pass a function that returns the activation function and the derivative. The problem is that activation and derivatives are not computed at the same time (meaning it would be slower) and also add a bit more complexity to code.

I meant passing two functions, but I'm happy with the private variable solution as well.

rth committed:
Don't publicaly re-export ACTIVATIONS variable from sklearn.neural_network._base

adrinjalali reviewed (approved):
Thanks @rth

rth commented:
> Are we going to start recommending users to do this in our issue board?

Allowing to do this via a private API implying that we are not officially supporting it and it can change in the future, is not such a bad compromise. But it still leaves people the ability to experiment if they are so inclined. Documenting it would make this mechanism a bit more public, I'm not sure that is something we want.

rth commented on sklearn/neural_network/multilayer_perceptron.py:
```diff
@@ -97,6 +97,10 @@ def _forward_pass(self, activations):
         """
+        # local import to allow custom activations set via this
+        # private variable in _base
+        from ._base import ACTIVATIONS
```
First to remove `ACTIVATIONS` and `DERIVATIVES` from a public import path. Second to allow monkeypatching it. If you import it at the top level, it's imported together with `MLPClassifier`, and then changing `neural_network._base.ACTIVATION` no longer have an effect on `MLPClassifier`.

rth replied:
You are right. Actually the above example will work even with top level import. The problem is pytest monkeypatching, if I don't do local imports tests using it will fail, since I monkeypatch `ACTIVATIONS` instead of just updating it to be sure not to affect the global state or other tests.

I removed the comment as it was not accurate, but I guess in any case it doesn't hurt not to expose those variables at top level?

rth replied:
> It's not great, but having a local import so that we can write a test in a certain way sounds like a strange design to me

Fair enough, I can also just revert https://github.com/scikit-learn/scikit-learn/pull/14815/commits/7519f1412eeb6ab99d5ba0dfdafc1f893642ca9a and we will be back where we started but with an example in the issue using the private API :)

rth committed:
Update comment

ArijitN5 commented:
Cutom 

> A minimal example would be something like,
> 
> ```python
> from sklearn.neural_network.multilayer_perceptron import MLPClassifier
> from sklearn.neural_network._base import ACTIVATIONS, DERIVATIVES                         
> 
> ACTIVATIONS['identity2'] = lambda x: x                                                                                                 
> 
> def inplace_identity_derivative2(Z, delta): 
>      pass 
>                                                                                                                                         
> DERIVATIVES['identity2'] = inplace_identity_derivative2                                                                                
> 
> mlp = MLPClassifier(activation='identity2')                                                                                            
> 
> mlp.fit([[0, 1], [1, 1]], [0, 1])
> ```
> 
> changing public variables in the scikit-learn namespace is not great, but as a somewhat experimental solution it should work (also it doesn't cost much).



This is not helping me to work with custom activation functions in Sklearn . 
Please suggest what process should I do.

adrinjalali commented:
> Could we consider (privately) accept a tuple of callables instead of hacking the dict?

I'd be happy to publicly accepting a tuple of callables, why not?

rth commented:
> I'd be happy to publicly accepting a tuple of callables, why not?

The problem is that tuple of callables doesn't generalize well to the case when more than 2 callable are needed (e.g. GLM) https://github.com/scikit-learn/scikit-learn/issues/14720#issuecomment-523938695 so it's not a general solution to this problem.

rth committed:
Revert "Update comment"

This reverts commit b649abb9578e818ca2aa0e13bed91090a0e6c026.

rth committed:
Revert "Don't publicaly re-export ACTIVATIONS variable from sklearn.neural_network._base"

This reverts commit 7519f1412eeb6ab99d5ba0dfdafc1f893642ca9a.

rth commented:
OK so I reverted latest changes, and can remove the test if it bothers you. So there would be only a 2 line code simplification PR. The example in the PR still uses the private API.

To summarize, if we were to do this as part of the public API, we could

 a) add a `activation_derivative=None` parameter, and accept a callable there and for `activation`. 
 b) have some registration mechanism for `ACTIVATION` (e.g. `register_activation('name', activation, derivative)`
 c) accept a tuple of callable with `(fn activation, fn derivative)`
 d) accept a callable returning a tuple
 e) Accept an instance (or a named tuple) of the form .


     class Activation:
        def __call___(...)
        
        def derivative(...)

Can't say I'm too enthusiastic about any of them. Putting all constraints aside I would probably go with e) or c) (with a named tuple) .

adrinjalali commented:
I'd be fine with not having the test, and I'm also fine with C or E. After these discussions I'm also just happy with the PR as is, but ideally I'd open an issue to have one of your suggestions implemented after this one.

rth committed:
Remove custom activations test

NicolasHug commented:
I'm OK with the PR as-is too.

If we ever write an example we might want to use `_base.ACTIVATIONS` instead of the public path.

rth commented:
I removed the test and changed the PR title. Should be good to go. 

Will try to open an issue about possible public API implementation later today.

NicolasHug commented:
Merging, thanks @rth 	

I'm OK with a public API but I think we decided not to add more features to the NN code during the meeting? So might just not be worth it
